session_connect_test.go

```diff
@@ -0,0 +1,82 @@
+package gocql
+
+import (
+	"log"
+	"net"
+	"testing"
+	"time"
+)
+
+func setupTestTcpServer(t *testing.T) (net.Listener, string, int) {
+	srvr, err := net.Listen("tcp4", "0.0.0.0:0")
+	assertNil(t, "error when creating tcp server", err)
+	testHost, testPort, err := parseHostPort(srvr.Addr().String())
+	log.Printf("using host=%s and port=%d for test tcp server (err=%v)", testHost, testPort, err)
+	return srvr, testHost, testPort
+}
+
+func tcpConnWaitAndClose(t *testing.T, srvr net.Listener, tchan chan<- error) {
+	conn, err := srvr.Accept()
+	if conn != nil {
+		log.Printf("connection received from addr=%+v", conn.RemoteAddr())
+		conn.Close()
+	}
+	tchan <- err
+}
+
+func testConnErrorHandler(t *testing.T) ConnErrorHandler {
+	return connErrorHandlerFn(func(conn *Conn, err error, closed bool) {
+		t.Errorf("in connection handler: %v", err)
+	})
+}
+
+func assertConnectionEventually(t *testing.T, delay time.Duration, tchan <-chan error) {
+	select {
+	case err := <-tchan:
+		if err != nil {
+			t.Errorf("accepting connection: %v", err)
+		}
+	case <-time.After(delay):
+		t.Error("timed out waiting for connection")
+	}
+}
+
+func TestSession_connect_WithNoTranslator(t *testing.T) {
+	srvr, testHost, testPort := setupTestTcpServer(t)
+	defer srvr.Close()
+
+	session := createTestSession()
+	defer session.Close()
+
+	tchan := make(chan error)
+	defer close(tchan)
+	go tcpConnWaitAndClose(t, srvr, tchan)
+
+	session.connect(srvr.Addr().String(), testConnErrorHandler(t), &HostInfo{
+		peer: testHost,
+		port: testPort,
+	})
+
+	assertConnectionEventually(t, 500*time.Millisecond, tchan)
+}
+
+func TestSession_connect_WithTranslator(t *testing.T) {
+	srvr, testHost, testPort := setupTestTcpServer(t)
+	defer srvr.Close()
+
+	session := createTestSession()
+	defer session.Close()
+	session.cfg.AddressTranslator = staticAddressTranslator(testHost, testPort)
+
+	tchan := make(chan error)
+	defer close(tchan)
```

Comment:
if we hit the timeout then return from this function, `tchan` will be closed and potentially trigger a send on closed channel.